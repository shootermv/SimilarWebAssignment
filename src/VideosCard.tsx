import * as React from 'react';
import { Card, List, Input, Button } from 'semantic-ui-react';
import {/*SortableContainer, SortableElement,*/ arrayMove } from 'react-sortable-hoc';

interface VideosCardProps extends React.HTMLProps<HTMLDivElement> {
  tags: any[];
  tag?: string;
  onTagAdded(e: any): void;
  onVideoRemoved(e: any, idx: number): void;
}

export default class VideosCard extends React.Component<VideosCardProps, any> {
  public state: any;
  public inpt: any;
  constructor(props: VideosCardProps) {
    super(props);
    this.state = { textInput: '', videos: props.tags };
  }
  componentWillReceiveProps(nextProps: any) {
    this.setState({ videos: nextProps.tags });
  }

  onSortEnd = ({ oldIndex, newIndex }: any) => {
    this.setState({
      videos: arrayMove(this.state.videos, oldIndex, newIndex),
    });
  }

  handleKeyDown(e: any) {
    if (e.keyCode === 13) { // Enter key
      this.addVideo(null);
    }
  }
  handleChange(e: any) {
    const { value } = e.target;
    this.setState({ textInput: value });
  }
  addVideo(e: any) {
    if (this.inpt.value.indexOf('www.youtube.com/watch?v=') === -1) {
      return;
    } // not valid string
    if (this.state.videos.filter((t: any) => this.inpt.value === t.fieldValue).length) {
      return;
    } // same video again  


    this.inpt.value = '';
    this.props.onTagAdded({ fieldValue: this.state.textInput, title: this.state.textInput });
  }
  removeVideo(e: any, idx: number) {
    this.props.onVideoRemoved(e, idx);
  }
  render() {
    const { videos } = this.state;

    return (
      <Card>
        <Card.Content>
          <Card.Header style={{ display: 'flex' }}>
            <Input placeholder="place here url of video..."
              name="textInput" onChange={(e) => this.handleChange(e)} onKeyDown={(e: any) => this.handleKeyDown(e)}>
              <input ref={(inpt) => this.inpt = inpt} />
            </Input>
            <Button icon="plus" onClick={(e) => this.addVideo(e)} style={{ margin: '0 0 0 5px' }} />
          </Card.Header>
        </Card.Content>
        <Card.Content>
          <List>
            {videos.map((video: any, idx: number) => {
              const isActive = video.fieldValue === this.props.tag;
              const activeStyle = {
                color: 'blue',
                fontWeight: '700',
              };
              return (
                <List.Item as="span" key={video.fieldValue}>
                  <List.Icon name="remove" onClick={(e: any) => this.removeVideo(e, idx)} />
                  <List.Content style={isActive ? activeStyle : null} >
                    {video.title} {video.duration}
                  </List.Content>
                </List.Item>
              );
            })}
          </List>
        </Card.Content>
      </Card>
    );
  }
}
