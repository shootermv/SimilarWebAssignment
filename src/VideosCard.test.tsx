import { shallow } from "enzyme";
import "jest";
import * as React from "react";
import VideosCard from "./VideosCard";

import { List } from "semantic-ui-react";


describe("VideosCard component", () => {


  it("should list all the tags", () => {
    const tags = [
      { fieldValue: "tag01", totalCount: 2 },
      { fieldValue: "tag02", totalCount: 4 },
      { fieldValue: "tag03", totalCount: 6 },
    ] as any[];

    const wrapper = shallow(<VideosCard
      tags={tags} tag={tags[0]}
      onTagAdded={() => console.log("")}
      onVideoRemoved={() => console.log("")} />);

    expect(wrapper.find(List.Item)).toHaveLength(3);
  });

  it("should have on tag active", () => {
    const tags = [
      { fieldValue: "tag01", totalCount: 2 },
      { fieldValue: "tag02", totalCount: 4 },
      { fieldValue: "tag03", totalCount: 6 },
    ] as any[];

    const wrapper = shallow(<VideosCard
      tags={tags} tag={tags[0]}
      onTagAdded={() => console.log("")}
      onVideoRemoved={() => console.log("")} />);

    expect(wrapper).toMatchSnapshot();
  });
});
