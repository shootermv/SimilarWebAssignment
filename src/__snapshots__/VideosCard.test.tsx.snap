// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideosCard component should have on tag active 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <Card>
    <CardContent>
        <CardHeader
            style={
                Object {
                    "display": "flex",
                  }
            }
        >
            <Input
                name="textInput"
                onChange={[Function]}
                onKeyDown={[Function]}
                placeholder="place here url of video..."
                type="text"
            >
                <input />
            </Input>
            <Button
                as="button"
                icon="plus"
                onClick={[Function]}
                style={
                    Object {
                        "margin": "0 0 0 5px",
                      }
                }
            />
        </CardHeader>
    </CardContent>
    <CardContent>
        <List>
            <ListItem
                as="span"
            >
                <ListIcon
                    name="remove"
                    onClick={[Function]}
                />
                <ListContent
                    style={
                        Object {
                            "color": "blue",
                            "fontWeight": "700",
                          }
                    }
                >
                    Curse Of Black Pearl
                     
                    1
                </ListContent>
            </ListItem>
            <ListItem
                as="span"
            >
                <ListIcon
                    name="remove"
                    onClick={[Function]}
                />
                <ListContent
                    style={null}
                >
                    Sleepy Hollow
                     
                    2
                </ListContent>
            </ListItem>
            <ListItem
                as="span"
            >
                <ListIcon
                    name="remove"
                    onClick={[Function]}
                />
                <ListContent
                    style={null}
                >
                    Eduward Scissorhands
                     
                    16
                </ListContent>
            </ListItem>
        </List>
    </CardContent>
</Card>,
  "nodes": Array [
    <Card>
      <CardContent>
            <CardHeader
                  style={
                        Object {
                              "display": "flex",
                            }
                  }
            >
                  <Input
                        name="textInput"
                        onChange={[Function]}
                        onKeyDown={[Function]}
                        placeholder="place here url of video..."
                        type="text"
                  >
                        <input />
                  </Input>
                  <Button
                        as="button"
                        icon="plus"
                        onClick={[Function]}
                        style={
                              Object {
                                    "margin": "0 0 0 5px",
                                  }
                        }
                  />
            </CardHeader>
      </CardContent>
      <CardContent>
            <List>
                  <ListItem
                        as="span"
                  >
                        <ListIcon
                              name="remove"
                              onClick={[Function]}
                        />
                        <ListContent
                              style={
                                    Object {
                                          "color": "blue",
                                          "fontWeight": "700",
                                        }
                              }
                        >
                              Curse Of Black Pearl
                               
                              1
                        </ListContent>
                  </ListItem>
                  <ListItem
                        as="span"
                  >
                        <ListIcon
                              name="remove"
                              onClick={[Function]}
                        />
                        <ListContent
                              style={null}
                        >
                              Sleepy Hollow
                               
                              2
                        </ListContent>
                  </ListItem>
                  <ListItem
                        as="span"
                  >
                        <ListIcon
                              name="remove"
                              onClick={[Function]}
                        />
                        <ListContent
                              style={null}
                        >
                              Eduward Scissorhands
                               
                              16
                        </ListContent>
                  </ListItem>
            </List>
      </CardContent>
</Card>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <VideosCard
        onTagAdded={[Function]}
        onVideoRemoved={[Function]}
        tag="tag01"
        videos={
                Array [
                        Object {
                          "duration": 1,
                          "fieldValue": "tag01",
                          "title": "Curse Of Black Pearl",
                        },
                        Object {
                          "duration": 2,
                          "fieldValue": "tag02",
                          "title": "Sleepy Hollow",
                        },
                        Object {
                          "duration": 16,
                          "fieldValue": "tag03",
                          "title": "Eduward Scissorhands",
                        },
                      ]
        }
/>,
      "_debugID": 3,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": VideosCard {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "onSortEnd": [Function],
        "props": Object {
          "onTagAdded": [Function],
          "onVideoRemoved": [Function],
          "tag": "tag01",
          "videos": Array [
            Object {
              "duration": 1,
              "fieldValue": "tag01",
              "title": "Curse Of Black Pearl",
            },
            Object {
              "duration": 2,
              "fieldValue": "tag02",
              "title": "Sleepy Hollow",
            },
            Object {
              "duration": 16,
              "fieldValue": "tag03",
              "title": "Eduward Scissorhands",
            },
          ],
        },
        "refs": Object {},
        "state": Object {
          "textInput": "",
          "videos": Array [
            Object {
              "duration": 1,
              "fieldValue": "tag01",
              "title": "Curse Of Black Pearl",
            },
            Object {
              "duration": 2,
              "fieldValue": "tag02",
              "title": "Sleepy Hollow",
            },
            Object {
              "duration": 16,
              "fieldValue": "tag03",
              "title": "Eduward Scissorhands",
            },
          ],
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 2,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <Card>
          <CardContent>
                    <CardHeader
                              style={
                                        Object {
                                                  "display": "flex",
                                                }
                              }
                    >
                              <Input
                                        name="textInput"
                                        onChange={[Function]}
                                        onKeyDown={[Function]}
                                        placeholder="place here url of video..."
                                        type="text"
                              >
                                        <input />
                              </Input>
                              <Button
                                        as="button"
                                        icon="plus"
                                        onClick={[Function]}
                                        style={
                                                  Object {
                                                            "margin": "0 0 0 5px",
                                                          }
                                        }
                              />
                    </CardHeader>
          </CardContent>
          <CardContent>
                    <List>
                              <ListItem
                                        as="span"
                              >
                                        <ListIcon
                                                  name="remove"
                                                  onClick={[Function]}
                                        />
                                        <ListContent
                                                  style={
                                                            Object {
                                                                      "color": "blue",
                                                                      "fontWeight": "700",
                                                                    }
                                                  }
                                        >
                                                  Curse Of Black Pearl
                                                   
                                                  1
                                        </ListContent>
                              </ListItem>
                              <ListItem
                                        as="span"
                              >
                                        <ListIcon
                                                  name="remove"
                                                  onClick={[Function]}
                                        />
                                        <ListContent
                                                  style={null}
                                        >
                                                  Sleepy Hollow
                                                   
                                                  2
                                        </ListContent>
                              </ListItem>
                              <ListItem
                                        as="span"
                              >
                                        <ListIcon
                                                  name="remove"
                                                  onClick={[Function]}
                                        />
                                        <ListContent
                                                  style={null}
                                        >
                                                  Eduward Scissorhands
                                                   
                                                  16
                                        </ListContent>
                              </ListItem>
                    </List>
          </CardContent>
</Card>,
        "_debugID": 4,
        "_renderedOutput": <Card>
          <CardContent>
                    <CardHeader
                              style={
                                        Object {
                                                  "display": "flex",
                                                }
                              }
                    >
                              <Input
                                        name="textInput"
                                        onChange={[Function]}
                                        onKeyDown={[Function]}
                                        placeholder="place here url of video..."
                                        type="text"
                              >
                                        <input />
                              </Input>
                              <Button
                                        as="button"
                                        icon="plus"
                                        onClick={[Function]}
                                        style={
                                                  Object {
                                                            "margin": "0 0 0 5px",
                                                          }
                                        }
                              />
                    </CardHeader>
          </CardContent>
          <CardContent>
                    <List>
                              <ListItem
                                        as="span"
                              >
                                        <ListIcon
                                                  name="remove"
                                                  onClick={[Function]}
                                        />
                                        <ListContent
                                                  style={
                                                            Object {
                                                                      "color": "blue",
                                                                      "fontWeight": "700",
                                                                    }
                                                  }
                                        >
                                                  Curse Of Black Pearl
                                                   
                                                  1
                                        </ListContent>
                              </ListItem>
                              <ListItem
                                        as="span"
                              >
                                        <ListIcon
                                                  name="remove"
                                                  onClick={[Function]}
                                        />
                                        <ListContent
                                                  style={null}
                                        >
                                                  Sleepy Hollow
                                                   
                                                  2
                                        </ListContent>
                              </ListItem>
                              <ListItem
                                        as="span"
                              >
                                        <ListIcon
                                                  name="remove"
                                                  onClick={[Function]}
                                        />
                                        <ListContent
                                                  style={null}
                                        >
                                                  Eduward Scissorhands
                                                   
                                                  16
                                        </ListContent>
                              </ListItem>
                    </List>
          </CardContent>
</Card>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <VideosCard
    onTagAdded={[Function]}
    onVideoRemoved={[Function]}
    tag="tag01"
    videos={
        Array [
            Object {
              "duration": 1,
              "fieldValue": "tag01",
              "title": "Curse Of Black Pearl",
            },
            Object {
              "duration": 2,
              "fieldValue": "tag02",
              "title": "Sleepy Hollow",
            },
            Object {
              "duration": 16,
              "fieldValue": "tag03",
              "title": "Eduward Scissorhands",
            },
          ]
    }
/>,
}
`;
